### ISSUE TITLE
Consider invalidating attribute narrowing based on method calls known to mutate that attribute

### ISSUE THREAD

**Issue by jbree** [astral-sh/ty#2792]
### Summary

I couldn't find this in the tracker, but it feels like it may be a duplicate.

`self.bar` is initialized to `None`. Then a method call changes its value to `3.14`. ty narrows the type to `None` even though it is in fact `float` at this point.

https://play.ty.dev/49740290-cb43-4229-9e07-af8eab00fec4

### Version

0.0.16

**Comment by carljm** [comment]
Yes, attribute narrowing can be unsound in this way. And yet all Python type checkers do it anyway (every Python type checker reports the same result in this example), because the consequences of not doing it seem to be even worse, in that many common sound programs are then rejected by the type checker. Accurately dividing the false negatives from the false positives requires quite expensive (even intractable) cross-procedural analysis.

That said, it might be feasible to catch some of the most common cases of this: the cases where a method on `self` is called directly, which mutates an attribute on `self`, and discard the narrowing on that attribute when we see such a call. Will leave this issue open to track the possibility of trying that in the future.

**Comment by carljm** [comment]
#885 discusses the same issue, but tracks another possible "solution" (an opt-in setting to disable this unsound narrowing). So I'll leave this open to track the possibility that we could be somewhat more precise.

**Comment by fredericmorin-flare** [comment]
Same issue here when using sqlalchemy:

```
assert user.pw_reset_link is None
# test some logic here
# then validate sql state update through sqlalchemy
session.refresh(user) # refresh is invisible to ty 
assert user.pw_reset_link # get ty: redundant-condition
```
